src/webgpu/api/validation/buffer/destroy.spec.ts

```diff
@@ -74,18 +75,26 @@ g.test('while_mapped')
           mapMode: GPUConst.MapMode.READ,
         },
       ])
-      .filter(p => p.mappedAtCreation === false && p.mapMode === undefined)
+      .unless(p => p.mappedAtCreation === false && p.mapMode === undefined)
   )
   .fn(async t => {
-    const { usage, mapMode, mappedAtCreation } = t.params;
+    const { usage, mapMode, mappedAtCreation, unmapBeforeDestoy } = t.params;
     const buf = t.device.createBuffer({
       size: 4,
       usage,
       mappedAtCreation,
     });
 
-    if (!mappedAtCreation && mapMode !== undefined) {
+    if (mappedAtCreation && unmapBeforeDestoy) {
+      buf.unmap();
+    }
+
+    if ((!mappedAtCreation || unmapBeforeDestoy) && mapMode !== undefined) {
       await buf.mapAsync(mapMode);
+
+      if (unmapBeforeDestoy) {
+        buf.unmap();
+      }
     }
```

Comment:
I had hard time following this logic, in particular the guarantee between `mappedAtCreation && unmapBeforeDestroy` and `!mappedAtCreation || unmapBeforeDestroy`. After working through the possibilities I am pretty sure it's right, but this seems like it would be more direct:

````suggestion
    if (mapMode !== undefined) {
      if (mappedAtCreation) {
        buf.unmap();
      }
      await buf.mapAsync(mapMode);
    }

    if (unmapBeforeDestroy) {
      buf.unmap();
    }
```

Reply:
In combination with your two new suggested cases, yes. This looks like it would cover it.